pkg/cluster/cluster.go

```diff
@@ -177,39 +177,50 @@ func New(log log.Logger, reg prometheus.Registerer, clusterEnabled bool, listenA
 		return nil, err
 	}
 
-	// Attempt to start the Node by connecting to the peers in gossipConfig.
-	// If we cannot connect to any peers, fall back to bootstrapping a new
-	// cluster by ourselves.
-	err = gossipNode.Start()
-	if err != nil {
-		level.Debug(log).Log("msg", "failed to connect to peers; bootstrapping a new cluster")
-		gossipConfig.JoinPeers = nil
-		err = gossipNode.Start()
+	return &Clusterer{Node: gossipNode}, nil
+}
+
+// Start starts the node.
+// For the localNode implementation, this is a no-op.
+// For the gossipNode implementation, Start will attempt to connect to the
+// configured list of peers; if this fails it will fall back to bootstrapping a
+// new cluster of its own.
+func (c *Clusterer) Start() error {
+	switch node := c.Node.(type) {
+	case *localNode:
+		return nil // no-op, always ready
+	case *GossipNode:
+		err := node.Start()
 		if err != nil {
-			return nil, err
+			level.Debug(node.log).Log("msg", "failed to connect to peers; bootstrapping a new cluster")
+			node.cfg.JoinPeers = nil
+			err = node.Start()
+			if err != nil {
+				return err
+			}
 		}
-	}
 
-	// Nodes initially join the cluster in the Viewer state. We can move to the
-	// Participant state to signal that we wish to participate in reading or
-	// writing data.
-	err = gossipNode.ChangeState(context.Background(), peer.StateParticipant)
-	if err != nil {
-		return nil, err
-	}
-
-	res := &Clusterer{Node: gossipNode}
-
-	gossipNode.Observe(ckit.FuncObserver(func(peers []peer.Peer) (reregister bool) {
-		names := make([]string, len(peers))
-		for i, p := range peers {
-			names[i] = p.Name
+		// We now have either joined or started a new cluster.
+		// Nodes initially join in the Viewer state. We can move to the
+		// Participant state to signal that we wish to participate in reading
+		// or writing data.
+		err = node.ChangeState(context.Background(), peer.StateParticipant)
```

Comment:
This will block until the message has been broadcast to at least one peer; cancelling the context here will still try to broadcast the message but we'll not wait for the confirmation.

Since joining a new cluster will prompt other peers to Update their components and re-distribute the work right away, I think we should validate that we've successfully broadcast this node's intention to participate in splitting the workload, too. WDYT?

Reply:
I was thinking more about the case where we'll wait forever for the state to change - e.g. when we try to broadcast to a single peer that became unresponsive. 

But yeah, I gather if we had a timeout here, we could broadcast state change to become a Participant successfully and at the same time timeout, leaving with an incorrect cluster state. Would the cluster still recover because there would be no heartbeats after this instance quits? 

Another question: after `ChangeState` returns - did at least one peer successfully receive the broadcast, or did we just successfully _send_ the broadcast, but don't know if it was received? I couldn't immediately see this from the code.